README.md
# Angular Tour - [AngularJS](http://angularjs.org/) directive for giving a tour of your website.

## Demo

Want to see it in action? Visit <http://daftmonk.github.io/angular-tour/>

## Showcase features of your website

Give an interactive tour to showcase the features of your website. 

  * Easy to use
  * Responsive to window resizes
  * Smoothly scrolls to each step
  * Control the placement for each tour tip

## Supported browsers

Has been tested in 

* Chrome
* Firefox
* Safari
* Internet Explorer 9+

## Installation

To install run

```zsh
bower install angular-tour
```

Angular Tour has a dependency on jQuery.

## Setup

Once bower has downloaded the dependencies for you, you'll need to make sure you add the required libraries to your index file. Your script includes should look something like this:

```HTML
<script src="bower_components/jquery/jquery.js"></script>
<script src="bower_components/angular/angular.js"></script>
<script src="bower_components/angular-tour/dist/angular-tour-tpls.min.js"></script>
```

You'll also probably want to include the default stylesheet for angular tour. (You can replace this with your own stylesheet.)

```HTML
<link href="bower_components/angular-tour/dist/angular-tour.css" rel="stylesheet" type="text/css"/>
```

Lastly, you'll need to include the module in your angular app

```JS
angular.module('myApp', ['angular-tour'])
```

## How to use

To begin your tour you'll need a `<tour>` element to contain all of your tour tips, it must have a `step` attribute for binding the tour step to your scope.

Add the tourtip attribute to whatever elements you want to add a tip to.

Example markup:

```HTML
<tour step="currentStep">
  <span tourtip="tip 1"> Highlighted </span>
  <span tourtip="tip 2"> Elements </span>
  <input tourtip="or add it as an attribute to your element" />
</tour>
```

You can also add callbacks to the `tour`:

```HTML
<tour step="currentStep" post-tour="tourEnded()" post-step="stepComplete()" tour-complete="tourComplete()">
```

* `tourEnded` will be called always when tour will be ended - completed or not
* `tourComplete` will be called only when user will get to the last step
* `stepComplete` will be called every time the step will be changed

> Side note: If you don't initialize `currentStep` in your controller it will be by default set to `-1`, which mean the tour won't appear on page load. 
> This is **breaking change**, as previously `currentStep` was defaulted to `0`, which caused tour to start automatically.

It is very easy to add a cookie module that remembers what step a user was on. Using the angular-cookie module this is all you need to integrate cookies:

```JS
// load cookie, or start new tour
$scope.currentStep = ipCookie('myTour') || 0;

// save cookie after each step
$scope.stepComplete = function() {
  ipCookie('myTour', $scope.currentStep, { expires: 3000 });
};
```
There are additional attributes that allow you to customize each tour-tip.

* `tourtip-step` **(Default: "null")**: tour tips play from step 0 onwards, or in the order they were added. You can specify a specific order, e.g.

```HTML
<span tourtip="tip 2" tourtip-step="1"></span>
<span tourtip="tip 1" tourtip-step="0"></span>
<span tourtip="tip 3" tourtip-step="2"></span>
```

* `tourtip-next-label` **(Default: "Next")**: The text for the next button.

* `tourtip-placement` **(Default: "top")**: Placement of the tour tip relative to the target element. can be top, right, left, bottom

* `on-show` **(Default: null)**: Callback, which will be called when the tour step will appear

* `on-proceed` **(Default: null)**: Callback, which will be called when user move to the next step, but just before showing it

* `tourtip-element` **(Default: null)**: CSS Selector for element, for which tourtip will be pointed. If left `null`, tourtip will be pointed for itself

* `use-source-scope` **(Default: false)**: Option meaningful only when using virtual steps. When set to `false` - it will use as a target scope, scope of target's element for evaluating `on-show` and `on-proceed` callbacks. When set to `true`, target's scope will be scope when resides step itself. See example below for a better explanation if you still need one.

* `tourtip-container-element` **(Default: "body")**: Element the tour tips are placed in. Especially practical when you have several scrollable containers you want to display tours in.

Inside your tour, you also have access to two scope methods for ending and starting the tour.

```HTML
<a ng-click="openTour()">Open Tour</a>
<a ng-click="closeTour()">Close Tour</a>
```

## Virtual steps

If you have more complicated structure of application, especially with page 
divided by page includes and different controllers you can consider using this approach.

```html
<div class="container">
  <a class="btn btn-sm magic-button" ng-click="someRandomAction()">Well, some magic button</a>
  <a id="other-button" class="btn btn-s">Well, some magic button</a>
</div>

<!-- somewhere else on the page, and different scope -->

<button class="btn btn-sm btn-primary" ng-click="localAction()">Pff</button>

<tour step="currentStep">
  <virtual-step 
    tourtip="Content of the first step"
    tourtip-next-label="Move forward"
    tourtip-placement="bottom"
    tourtip-element=".magic-button"
    on-show="someRandomAction()"
    tourtip-container-element="#scrollableDiv"
    tourtip-step="0" />  
  <div
    tourtip="Some other content..."
    tourtip-next-label="Faster, faster!"
    tourtip-placement="top"
    tourtip-element="#other-button"
    on-proceed="localAction()"
    use-source-scope="true"
    tourtip-container-element="#scrollableDiv"
    tourtip-step="1" />
</tour>
```

Name of the tag doesn't really matter. It's a normal step definition, but the element that will be used to attach to is specified by `virtual-step` attribute.

## Customization

### Defaults

If you'd like to edit the defaults for all your tour, you can inject tourConfig somewhere into your app and modify the following defaults.

```JS
{
  placement        : 'top',  // default placement relative to target. 'top', 'right', 'left', 'bottom', 'center', 'center-top'
  animation        : true,   // if tips fade in
  nextLabel        : 'Next', // default text in the next tip button
  scrollSpeed      : 500,    // page scrolling speed in milliseconds
  margin           : 28,     // margin in pixels that the tip is from the target (matches placement)
  backDrop         : false,  // should page dim out when the tour starts?
  containerElement : 'body'  // default container element to parent tourtips to
}
```

### Positioning

Tourtip positioning can be controlled globally in the previously mentioned tourConfig service via the placement property ('top', 'right', 'bottom', 'center', 'center-top'), or via the tourtip-placement attribute, which will allow you to set the placement on an individual element-by-element basis. 

The distance between a tourtip and the element it is attached to can either be set globally via tourConfig.margin, or on an individual element-by-element basis using the tourtip-margin attribute. The margin will always match the placement - if the placement is top, tourtip-margin will add a margin between the tourtip and the top of the element. 

There may be times, especially when transcluding or applying to conditional elements, where the tour tip's calculated x,y position at compilation might not correspond to the element's current position. In these cases you can use tourtip-offset-horizontal or tourtip-offset-vertical to override and adjust the positioning by a certain amount of pixels.

```html

<tour step="currentStep">
  <p
    tourtip="Hey! I'd like to walk you through our site, it's great"
    tourtip-next-label="Hmmm, okay sure!"
    tourtip-placement="top"
    tourtip-margin="10"
    tourtip-step="0">
    Hi! Welcome to our site thing.
  </p>
  <p
    ng-show="currentStep === 1"
    tourtip="Behold! I am now explaining the feature..."
    tourtip-next-label="Wow, Amazing!"
    tourtip-placement="right"
    tourtip-offset-vertical="-30"
    tourtip-offset-horizontal="-26"
    tourtip-step="1" />
    Some cool feature... sure does need some splaining tho, dang...
  </p>
</tour>
```

### Customizing Templates

As was already mentioned, you can use your own CSS for styling the tour tips. You can also use your own markup.

If you would like to replace the html template, instead of using the `angular-tour-tpls.min.js` script, use `angular-tour.min.js` which doesn't include a template.

The easiest way to add your own template is to use the script directive:

```HTML
<script id="tour/tour.tpl.html" type="text/ng-template">
  <div class="tour-tip">
      <span class="tour-arrow tt-{{ ttPlacement }}" ng-hide="centered"></span>
      <div class="tour-content-wrapper">
          <p ng-bind="ttContent"></p>
          <a ng-click="proceed()" ng-bind="ttNextLabel" class="small button tour-next-tip"></a>
          <a ng-click="closeTour()" class="tour-close-tip">×</a>
      </div>
  </div>
</script>
```

## License

This project is licensed under the [MIT license](http://opensource.org/licenses/MIT).


CHANGELOG.md
<a name="0.2.2"></a>
# 0.2.2 (2015-07-22)

## Bugfix

- **Vertical and Horizontal Offsets**: properly initialize horizontal and vertical offsets if not specified

<a name="0.2.1"></a>
# 0.2.1 (2015-07-22)

## Features

- **Vertical and Horizontal Offsets**: allows fine-grained control of `tourtip` positioning - #39

## Breaking Changes

- `offset` property has been renamed to `margin` as not to cause confusion with (optional) vertical and horizontal positioning offset properties



<a name="0.2.0"></a>
# 0.2.0 (2015-07-22)

## Features

- **Miscellaneous Improvements**: Virtual Steps, `tourComplete` callback, additional tourtip callbacks (`on-show`, `on-proceed`), tourtip container targeting, optional backdrop, `center` and `center-top` options - #28

## Breaking Changes

- `tourtip`s are now appended to the `body` element by default

karma.conf.js
'use strict';

module.exports = function (config) {
  config.set({
    // base path, that will be used to resolve files and exclude
    basePath : './',

    // testing framework to use (jasmine/mocha/qunit/...)
    frameworks: ['jasmine'],

    files : [
        'bower_components/jquery/jquery.js',
        'bower_components/angular/angular.js',
        'bower_components/angular-cookie/angular-cookie.js',
        'bower_components/angular-mocks/angular-mocks.js',
        'src/**/*.js',
        'src/**/*.spec.js',
        'src/**/*.tpl.html'
    ],

    // location of templates
    preprocessors: {
        'src/**/*.tpl.html': 'html2js'
    },

    ngHtml2JsPreprocessor: {
        stripPrefix: 'src/'
    },

    // list of files / patterns to exclude
    exclude: [],

    reporters: ['progress'],

    colors: true,

    // web server port
    port: 9877,

    // level of logging
    // possible values: LOG_DISABLE || LOG_ERROR || LOG_WARN || LOG_INFO || LOG_DEBUG
    logLevel: config.LOG_INFO,

    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: false,

    // Start these browsers, currently available:
    // - Chrome
    // - ChromeCanary
    // - Firefox
    // - Opera
    // - Safari (only Mac)
    // - PhantomJS
    // - IE (only Windows)
    browsers: ['Chrome'],

    // Continuous Integration mode
    // if true, it capture browsers, run tests and exit
    singleRun : false
  });
};

demo/DemoApp.js
'use strict';

angular.module('DemoApp', ['ngRoute', 'angular-tour'])
  .config(function ($routeProvider) {
    $routeProvider
      .when('/', {
        templateUrl: 'main.html',
        controller: 'MainCtrl'
      })
      .when('/features', {
        templateUrl: 'features.html',
        controller: 'FeaturesCtrl'
      })
      .when('/example', {
        templateUrl: 'example.html',
        controller: 'ExampleCtrl'
      })
      .when('/howtoinstall', {
        templateUrl: 'howtoinstall.html',
        controller: 'HowToInstallCtrl'
      })
      .when('/github', {
        templateUrl: 'github.html',
        controller: 'GitHubCtrl'
      })
      .otherwise({
        redirectTo: '/'
      });
  }).controller('DemoCtrl', function($scope, $timeout, $location, tourState) {
    tourState.started();
    $scope.showTour = function() {
      $location.path('/');
      $scope.$broadcast('tourshow');
    };

    $scope.init = function() {
      $scope.$on('onTourEnd', function() {
        $location.path('/features');
      });
    };
  }).controller('FeaturesCtrl', function($scope, $location) {
    $scope.init = function() {
      $scope.$on('onTourEnd', function() {
        $location.path('/example');
      });
    }
    
  }).controller('ExampleCtrl', function($scope, $location) {
    $scope.init = function() {
      $scope.$on('onTourEnd', function() {
        $location.path('/howtoinstall');
      });
    }

  }).controller('HowToInstallCtrl', function($scope, $location) {
    $scope.init = function() {
      $scope.$on('onTourEnd', function() {
        $location.path('/github');
      });
    }
    
  }).controller('GitHubCtrl', function($scope, $location) {
    $scope.init = function() {
      $scope.$on('onTourEnd', function() {
        $location.path('/');
      });
    }
  }).controller('MainCtrl', function($scope) {
  });


src/tour/tour.js
'use strict';

angular.module('angular-tour.tour', [])

/**
 * tourConfig
 * Default configuration, can be customized by injecting tourConfig into your app and modifying it
 */
.constant('tourConfig', {
    placement: 'top', // default placement relative to target. 'top', 'right', 'left', 'bottom'
    animation: true, // if tips fade in
    nextLabel: 'Next', // default text in the next tip button
    scrollSpeed: 500, // page scrolling speed in milliseconds
    margin: 28, // how many pixels margin the tip is from the target
    backDrop: false, // if there is a backdrop (gray overlay) when tour starts
    useSourceScope: false, // only target scope should be used (only when using virtual steps)
    containerElement: 'body' // default container element to parent tourtips to
})

/**
 * TourController
 * the logic for the tour, which manages all the steps
 */
.controller('TourController', function($scope, orderedList) {
    var self = this,
        steps = self.steps = orderedList();

    // we'll pass these in from the directive
    self.postTourCallback = angular.noop;
    self.postStepCallback = angular.noop;
    self.showStepCallback = angular.noop;
    self.currentStep = -1;

    // if currentStep changes, select the new step
    $scope.$watch(function() {
            return self.currentStep;
        },
        function(val) {
            self.select(val);
        }
    );

    self.select = function(nextIndex) {
        if (!angular.isNumber(nextIndex)) return;

        self.unselectAllSteps();
        var step = steps.get(nextIndex);
        if (step) {
            step.ttOpen = true;
        }

        // update currentStep if we manually selected this index
        if (self.currentStep !== nextIndex) {
            self.currentStep = nextIndex;
        }

        if (self.currentStep > -1)
            self.showStepCallback();

        if (nextIndex >= steps.getCount()) {
            self.postTourCallback(true);
        }
        self.postStepCallback();
    };

    self.addStep = function(step) {
        if (angular.isNumber(step.index) && !isNaN(step.index)) {
            steps.set(step.index, step);
        } else {
            steps.push(step);
        }
    };

    self.unselectAllSteps = function() {
        steps.forEach(function(step) {
            step.ttOpen = false;
        });
    };

    self.cancelTour = function() {
        self.unselectAllSteps();
        self.postTourCallback(false);
    };

    $scope.openTour = function() {
        // open at first step if we've already finished tour
        var startStep = self.currentStep >= steps.getCount() || self.currentStep < 0 ? 0 : self.currentStep;
        self.select(startStep);
    };

    $scope.closeTour = function() {
        self.cancelTour();
    };
})

/**
 * Tour
 * directive that allows you to control the tour
 */
.directive('tour', function($parse, $timeout, tourConfig) {
    return {
        controller: 'TourController',
        restrict: 'EA',
        scope: true,
        link: function(scope, element, attrs, ctrl) {
            if (!angular.isDefined(attrs.step)) {
                throw ('The <tour> directive requires a `step` attribute to bind the current step to.');
            }
            var model = $parse(attrs.step);
            var backDrop = false;

            // Watch current step view model and update locally
            scope.$watch(attrs.step, function(newVal) {
                ctrl.currentStep = newVal;
            });

            ctrl.postTourCallback = function(completed) {
                angular.element('.tour-backdrop').remove();
                backDrop = false;
                angular.element('.tour-element-active').removeClass('tour-element-active');

                if (completed && angular.isDefined(attrs.tourComplete)) {
                    scope.$parent.$eval(attrs.tourComplete);
                }
                if (angular.isDefined(attrs.postTour)) {
                    scope.$parent.$eval(attrs.postTour);
                }
            };

            ctrl.postStepCallback = function() {
                if (angular.isDefined(attrs.postStep)) {
                    scope.$parent.$eval(attrs.postStep);
                }
            };

            ctrl.showStepCallback = function() {
                if (tourConfig.backDrop) {
                    angular.element(tourConfig.containerElement).append(angular.element('<div class="tour-backdrop"></div>'));

                    $timeout(function() {
                        $('.tour-backdrop').remove();
                        angular.element('<div class="tour-backdrop"></div>').insertBefore('.tour-tip');
                    }, 1000)

                    backDrop = true;
                }
            };

            // update the current step in the view as well as in our controller
            scope.setCurrentStep = function(val) {
                model.assign(scope.$parent, val);
                ctrl.currentStep = val;
            };

            scope.getCurrentStep = function() {
                return ctrl.currentStep;
            };
        }
    };
})

/**
 * Tourtip
 * tourtip manages the state of the tour-popup directive
 */
.directive('tourtip', function($window, $compile, $interpolate, $timeout, scrollTo, tourConfig, debounce) {
    var startSym = $interpolate.startSymbol(),
        endSym = $interpolate.endSymbol();

    var template = '<div tour-popup></div>';

    return {
        require: '^tour',
        restrict: 'EA',
        scope: true,
        link: function(scope, element, attrs, tourCtrl) {
            attrs.$observe('tourtip', function(val) {
                scope.ttContent = val;
            });

            //defaults: tourConfig.placement
            attrs.$observe('tourtipPlacement', function(val) {
                scope.ttPlacement = (val || tourConfig.placement).toLowerCase().trim();
                scope.centered = (scope.ttPlacement.indexOf('center') === 0);
            });

            attrs.$observe('tourtipNextLabel', function(val) {
                scope.ttNextLabel = val || tourConfig.nextLabel;
            });

            attrs.$observe('tourtipContainerElement', function(val) {
                scope.ttContainerElement = val || tourConfig.containerElement;
            });

            attrs.$observe('tourtipMargin', function(val) {
                scope.ttMargin = parseInt(val, 10) || tourConfig.margin;
            });

            attrs.$observe('tourtipOffsetVertical', function(val) {
                scope.offsetVertical = parseInt(val, 10) || 0;
            });

            attrs.$observe('tourtipOffsetHorizontal', function(val) {
                scope.offsetHorizontal = parseInt(val, 10) || 0;
            });

            //defaults: null
            attrs.$observe('onShow', function(val) {
                scope.onStepShow = val || null;
            });

            //defaults: null
            attrs.$observe('onProceed', function(val) {
                scope.onStepProceed = val || null;
            });

            //defaults: null
            attrs.$observe('tourtipElement', function(val) {
                scope.ttElement = val || null;
            });

            //defaults: tourConfig.useSourceScope
            attrs.$observe('useSourceScope', function(val) {
                scope.ttSourceScope = !val ? tourConfig.useSourceScope : val === 'true';
            });

            //Init assignments (fix for Angular 1.3+)
            scope.ttNextLabel = tourConfig.nextLabel;
            scope.ttContainerElement = tourConfig.containerElement;
            scope.ttPlacement = tourConfig.placement.toLowerCase().trim();
            scope.centered = false;
            scope.ttMargin = tourConfig.margin;
            scope.offsetHorizontal = 0;
            scope.offsetVertical = 0;
            scope.ttSourceScope = tourConfig.useSourceScope;
            scope.ttOpen = false;
            scope.ttAnimation = tourConfig.animation;
            scope.index = parseInt(attrs.tourtipStep, 10);

            var tourtip = $compile(template)(scope);
            tourCtrl.addStep(scope);

            // wrap this in a time out because the tourtip won't compile right away
            $timeout(function() {
                scope.$watch('ttOpen', function(val) {
                    if (val) {
                        show();
                    } else {
                        hide();
                    }
                });
            }, 500);


            //determining target scope. It's used only when using virtual steps and there
            //is some action performed like on-show or on-progress. Without virtual steps
            //action would performed on element's scope and that would work just fine
            //however, when using virtual steps, whose steps can be placed in different
            //controller, so it affects scope, which will be used to run this action against.
            function getTargetScope() {
                var targetElement = scope.ttElement ? angular.element(scope.ttElement) : element;

                var targetScope = scope;
                if (targetElement !== element && !scope.ttSourceScope)
                    targetScope = targetElement.scope();

                return targetScope;
            }

            function calculatePosition(element, container) {
                var minimumLeft = 0; // minimum left position of tour tip
                var restrictRight;
                var ttPosition;

                // Get the position of the directive element
                var position = element[0].getBoundingClientRect();

                //make it relative against page or fixed container, not the window
                var top = position.top + window.pageYOffset;
                var containerLeft = 0;
                if (container && container[0]) {
                    top = top - container[0].getBoundingClientRect().top + container[0].scrollTop;
                    // if container is fixed, position tour tip relative to fixed container
                    if (container.css('position') === 'fixed') {
                        containerLeft = container[0].getBoundingClientRect().left;
                    }
                    // restrict right position if the tourtip doesn't fit in the container
                    var containerWidth = container[0].getBoundingClientRect().width;
                    if (tourtip.width() + position.width > containerWidth) {
                        restrictRight = containerWidth - position.left + scope.ttMargin;
                    }
                }

                var ttWidth = tourtip.width();
                var ttHeight = tourtip.height();

                // Calculate the tourtip's top and left coordinates to center it
                switch (scope.ttPlacement) {
                    case 'right':
                        var _left = position.left - containerLeft + position.width + scope.ttMargin + scope.offsetHorizontal;
                        ttPosition = {
                            top: top + scope.offsetVertical,
                            left: _left > 0 ? _left : minimumLeft
                        };
                        break;
                    case 'bottom':
                        var _left = position.left - containerLeft + scope.offsetHorizontal;
                        ttPosition = {
                            top: top + position.height + scope.ttMargin + scope.offsetVertical,
                            left: _left > 0 ? _left : minimumLeft
                        };
                        break;
                    case 'center':
                        var _left = position.left - containerLeft + 0.5 * (position.width - ttWidth) + scope.offsetHorizontal;
                        ttPosition = {
                            top: top + 0.5 * (position.height - ttHeight) + scope.ttMargin + scope.offsetVertical,
                            left: _left > 0 ? _left : minimumLeft
                        };
                        break;
                    case 'center-top':
                        var _left = position.left - containerLeft + 0.5 * (position.width - ttWidth) + scope.offsetHorizontal;
                        ttPosition = {
                            top: top + 0.1 * (position.height - ttHeight) + scope.ttMargin + scope.offsetVertical,
                            left: _left > 0 ? _left : minimumLeft
                        };
                        break;
                    case 'left':
                        var _left = position.left - containerLeft - ttWidth - scope.ttMargin + scope.offsetHorizontal;
                        ttPosition = {
                            top: top + scope.offsetVertical,
                            left: _left > 0 ? _left : minimumLeft,
                            right: restrictRight
                        };
                        break;
                    default:
                        var _left = position.left - containerLeft + scope.offsetHorizontal;
                        ttPosition = {
                            top: top - ttHeight - scope.ttMargin + scope.offsetVertical,
                            left: _left > 0 ? _left : minimumLeft
                        };
                        break;
                }

                ttPosition.top += 'px';
                ttPosition.left += 'px';

                return ttPosition;
            }

            function show() {
                if (!scope.ttContent) {
                    return;
                }

                if (scope.ttAnimation)
                    tourtip.fadeIn();
                else {
                    tourtip.css({
                        display: 'block'
                    });
                }

                var targetElement = scope.ttElement ? angular.element(scope.ttElement) : element;

                if (targetElement == null || targetElement.length === 0)
                    throw 'Target element could not be found. Selector: ' + scope.ttElement;

                angular.element(scope.ttContainerElement).append(tourtip);

                var updatePosition = function() {

                    var offsetElement = scope.ttContainerElement === 'body' ? undefined : angular.element(scope.ttContainerElement);
                    var ttPosition = calculatePosition(targetElement, offsetElement);

                    // Now set the calculated positioning.
                    tourtip.css(ttPosition);

                    // Scroll to the tour tip
                    var ttPositionTop = parseInt(ttPosition.top),
                        ttPositionLeft = parseInt(ttPosition.left);
                    scrollTo(tourtip, scope.ttContainerElement, -150, -300, tourConfig.scrollSpeed, ttPositionTop, ttPositionLeft);
                };

                if (tourConfig.backDrop)
                    focusActiveElement(targetElement);

                angular.element($window).bind('resize.' + scope.$id, debounce(updatePosition, 50));

                updatePosition();

                if (scope.onStepShow) {
                    var targetScope = getTargetScope();

                    //fancy! Let's make on show action not instantly, but after a small delay
                    $timeout(function() {
                        targetScope.$eval(scope.onStepShow);
                    }, 300);
                }
            }

            function hide() {
                tourtip.detach();
                angular.element($window).unbind('resize.' + scope.$id);
            }

            function focusActiveElement(el) {
                angular.element('.tour-element-active').removeClass('tour-element-active');

                if (!scope.centered)
                    el.addClass('tour-element-active');
            }

            // Make sure tooltip is destroyed and removed.
            scope.$on('$destroy', function onDestroyTourtip() {
                angular.element($window).unbind('resize.' + scope.$id);
                tourtip.remove();
                tourtip = null;
            });

            scope.proceed = function() {
                if (scope.onStepProceed) {
                    var targetScope = getTargetScope();

                    $timeout(function() {
                        targetScope.$eval(scope.onStepProceed);
                    }, 100);
                }

                scope.setCurrentStep(scope.getCurrentStep() + 1);
            };
        }
    };
})

/**
 * TourPopup
 * the directive that actually has the template for the tip
 */
.directive('tourPopup', function() {
    return {
        replace: true,
        templateUrl: 'tour/tour.tpl.html',
        scope: true,
        restrict: 'EA',
        link: function(scope, element, attrs) {}
    };
})

/**
 * OrderedList
 * Used for keeping steps in order
 */
.factory('orderedList', function() {
    var OrderedList = function() {
        this.map = {};
        this._array = [];
    };

    OrderedList.prototype.set = function(key, value) {
        if (!angular.isNumber(key))
            return;
        if (key in this.map) {
            this.map[key] = value;
        } else {
            if (key < this._array.length) {
                var insertIndex = key - 1 > 0 ? key - 1 : 0;
                this._array.splice(insertIndex, 0, key);
            } else {
                this._array.push(key);
            }
            this.map[key] = value;
            this._array.sort(function(a, b) {
                return a - b;
            });
        }
    };
    OrderedList.prototype.indexOf = function(value) {
        for (var prop in this.map) {
            if (this.map.hasOwnProperty(prop)) {
                if (this.map[prop] === value)
                    return Number(prop);
            }
        }
    };
    OrderedList.prototype.push = function(value) {
        var key = this._array[this._array.length - 1] + 1 || 0;
        this._array.push(key);
        this.map[key] = value;
        this._array.sort(function(a, b) {
            return a - b;
        });
    };
    OrderedList.prototype.remove = function(key) {
        var index = this._array.indexOf(key);
        if (index === -1) {
            throw new Error('key does not exist');
        }
        this._array.splice(index, 1);
        delete this.map[key];
    };
    OrderedList.prototype.get = function(key) {
        return this.map[key];
    };
    OrderedList.prototype.getCount = function() {
        return this._array.length;
    };
    OrderedList.prototype.forEach = function(f) {
        var key, value;
        for (var i = 0; i < this._array.length; i++) {
            key = this._array[i];
            value = this.map[key];
            f(value, key);
        }
    };
    OrderedList.prototype.first = function() {
        var key, value;
        key = this._array[0];
        value = this.map[key];
        return value;
    };

    var orderedListFactory = function() {
        return new OrderedList();
    };

    return orderedListFactory;
})

/**
 * ScrollTo
 * Smoothly scroll to a dom element
 */
.factory('scrollTo', function() {
    return function(target, containerElement, offsetY, offsetX, speed, ttPositionTop, ttPositionLeft) {
        if (target) {
            offsetY = offsetY || -100;
            offsetX = offsetX || -100;
            speed = speed || 500;
            $('html,' + containerElement).stop().animate({
                scrollTop: ttPositionTop + offsetY,
                scrollLeft: ttPositionLeft + offsetX
            }, speed);
        } else {
            $('html,' + containerElement).stop().animate({
                scrollTop: 0
            }, speed);
        }
    };
})
.factory('debounce', function($timeout, $q) {
  return function(func, wait, immediate) {
    var timeout;
    var deferred = $q.defer();
    return function() {
      var context = this, args = arguments;
      var later = function() {
        timeout = null;
        if(!immediate) {
          deferred.resolve(func.apply(context, args));
          deferred = $q.defer();
        }
      };
      var callNow = immediate && !timeout;
      if ( timeout ) {
        $timeout.cancel(timeout);
      }
      timeout = $timeout(later, wait);
      if (callNow) {
        deferred.resolve(func.apply(context,args));
        deferred = $q.defer();
      }
      return deferred.promise;
    };
  };
});
